Add exceptional return to func_invoke in embedding doc (#268)
WebAssembly/exception-handling: document/core/appendix/embedding.rst

.. _embed-error:

Exceptions and Errors
~~~~~~~~~~~~~~~~~~~~~

> Invoking an exported function may throw or propagate exceptions, expressed by an auxiliary syntactic class:

Reviewer: Yeah, this is still needed, but instead of tagaddr and values I think it now needs to carry an exnaddr.

.. math::
   \begin{array}{llll}
   \production{exception} & \exception &::=& \ETHROW ~ \exnaddr \\
   \end{array}

> The exception instance :math:`exnaddr` identifies the exception thrown.

Reviewer: ```suggestion
The exception address :math:`exnaddr` identifies the exception thrown.
```

.. _embed-error:

Exceptions and Errors
~~~~~~~~~~~~~~~~~~~~~

> Invoke operations may throw or propagate exceptions, indicated by an auxiliary syntactic class:

Reviewer: ```suggestion
Invoking an exported function may throw or propagate exceptions, expressed by an auxiliary syntactic class:
```

.. math::
   \begin{array}{llll}
   \production{exception} & \exception &::=& \ETHROW ~ \tagaddr ~ val^\ast \\
   \end{array}

> The tag instance :math:`tagaddr` identifies the :ref:`type <syntax-tagtype>` of exception thrown. The values :math:`val^\ast` are the exception's payload; their types match the tag type's parameters.

Reviewer: The tagaddr identifies the exception tag itself. That _has_ a type, but multiple tags can have the same tagtype.
```suggestion
The tag instance :math:`tagaddr` identifies the exception thrown. The values :math:`val^\ast` are the exception's payload; their types match the tag type's parameters.
```

> Failure of an interface operation is indicated by an auxiliary syntactic class:

Reviewer: ```suggestion
Failure of an interface operation is also indicated by an auxiliary syntactic class:
```

Exceptions and Errors
~~~~~~~~~~~~~~~~~~~~~

Invoking an exported function may throw or propagate exceptions, expressed by an auxiliary syntactic class:

.. math::
   \begin{array}{llll}

>    \production{exception} & \exception &::=& \ETHROW ~ \tagaddr ~ val^\ast \\

Reviewer: ```suggestion
   \production{exception} & \exception &::=& \ETHROW ~ \exnaddr \\
```

   \end{array}

> The tag instance :math:`tagaddr` identifies the exception thrown. The values :math:`val^\ast` are the exception's payload; their types match the tag type's parameters.

Reviewer: This also needs updating accordingly.

   Implementations can refine it to carry suitable classifications and diagnostic messages.

.. _embed-exception:

Exceptions
~~~~~~~~~~

Invoke operations may throw or propagate exceptions, indicated by an auxiliary syntactic class:

.. math::
   \begin{array}{llllll}
   \production{exception} & \exception &::=& \ETHROW & exnaddr & val^\ast \\
   \end{array}

> 

Author: Do we need to add language here describing what the tag address and payload values mean or is just naming them like this sufficient?

.. index:: invocation, value, result
.. _embed-func-invoke:

> :math:`\F{func\_invoke}(\store, \funcaddr, \val^\ast) : (\store, \val^\ast ~|~ \error ~|~ \exception)`

Reviewer 2: @dschuff It looks the order of `exception` and `error` hasn't changed as suggested

> :math:`\F{func\_invoke}(\store, \funcaddr, \val^\ast) : (\store, \val^\ast ~|~ \error ~|~ \exception)`

Author: ah yes, fixed, thanks

......................................................................................................

1. Try :ref:`invoking <exec-invocation>` the function :math:`\funcaddr` in :math:`\store` with :ref:`values <syntax-val>` :math:`\val^\ast` as arguments:

  a. If it succeeds with :ref:`values <syntax-val>` :math:`{\val'}^\ast` as results, then let :math:`\X{result}` be :math:`{\val'}^\ast`.

>   b. Else if the result is a :ref:`throw context <syntax-throw-contexts>` for an uncaught exception with :ref:`tag <syntax-tag-instances>` :math:`{\tag}` and payload :ref:`values <syntax-val>` :math:`{\val'}^\ast`,  let :math:`\X{result}` be :math:`{tag'}`, :math:`{\val'}`.

Reviewer: ```suggestion
  b. Else, if the outcome is an exception with :ref:`tag <syntax-tagaddr>` :math:`\tagaddr` and payload :ref:`values <syntax-val>` :math:`{\val'}^\ast`,  let :math:`\X{result}` be :math:`\ETHROW~\tagaddr`~{\val'}^\ast`.
```

.. index:: invocation, value, result
.. _embed-func-invoke:

:math:`\F{func\_invoke}(\store, \funcaddr, \val^\ast) : (\store, \val^\ast ~|~ \exception ~|~ \error)`
......................................................................................................

1. Try :ref:`invoking <exec-invocation>` the function :math:`\funcaddr` in :math:`\store` with :ref:`values <syntax-val>` :math:`\val^\ast` as arguments:

  a. If it succeeds with :ref:`values <syntax-val>` :math:`{\val'}^\ast` as results, then let :math:`\X{result}` be :math:`{\val'}^\ast`.

  b. Else if the outcome is an exception with a thrown :ref:`exception <exec-throw_ref>` :math:`\REFEXNADDR~\exnaddr` as the result, then let :math:`\X{result}` be :math:`\ETHROW~\exnaddr`

  c. Else it has trapped, hence let :math:`\X{result}` be :math:`\ERROR`.

2. Return the new store paired with :math:`\X{result}`.

.. math::
   ~ \\
   \begin{array}{lclll}
   \F{func\_invoke}(S, a, v^\ast) &=& (S', {v'}^\ast) && (\iff \invoke(S, a, v^\ast) \stepto^\ast S'; F; {v'}^\ast) \\

>    \F{func\_invoke}(S, a, v^\ast) &=& (S', \ETHROW~a') && (\iff \invoke(S, a, v^\ast) \stepto^\ast S'; F; \XT[(\REFEXNADDR~\exnaddr)~\THROWREF] \\

Reviewer: ```suggestion
   \F{func\_invoke}(S, a, v^\ast) &=& (S', \ETHROW~a') && (\iff \invoke(S, a, v^\ast) \stepto^\ast S'; F; \XT[(\REFEXNADDR~a')~\THROWREF] \\
```

  a. If it succeeds with :ref:`values <syntax-val>` :math:`{\val'}^\ast` as results, then let :math:`\X{result}` be :math:`{\val'}^\ast`.

>   b. Else if the outcome is an exception with a thrown :ref:`exception <exec-throw_ref>` :math:`\REFEXNADDR~\exnaddr` as the result, then let :math:`\X{result}` be :math:`\exnaddr`

Reviewer: ```suggestion
  b. Else if the outcome is an exception with a thrown :ref:`exception <exec-throw_ref>` :math:`\REFEXNADDR~\exnaddr` as the result, then let :math:`\X{result}` be :math:`\ETHROW~\exnaddr`
```

   \end{array}

.. _embed-tag-type:

:math:`\F{tag\_type}(\store, \tagaddr) : \tagtype`
........................................................

1. Return :math:`S.\STAGS[a].\TAGITYPE`.

2. Post-condition: the returned :ref:`tag type <syntax-tagtype>` is :ref:`valid  <valid-tagtype>`.

.. math::
   \begin{array}{lclll}
   \F{tag\_type}(S, a) &=& S.\STAGS[a].\TAGITYPE \\
   \end{array}

.. index:: exception, exception address, store, exception instance, exception type
.. _embed-exception:

Exceptions
~~~~~~~~~~

.. _embed-exception-alloc:

:math:`\F{exception\_alloc}(\store, \tagaddr, \val) : (\store, \exnaddr)`
............................................................................

1. Pre-condition: :math:`\tagaddr` is an allocated :ref:`tag address <syntax-tagaddr>`.

2. Let :math:`\exnaddr` be the result of :ref:`allocating an exception <alloc-exception>` in :math:`\store` with :ref:`tag address <syntax-tagaddr>` :math:`\tagaddr` and initialization values :math:`\val^\ast`.

3. Return the new store paired with :math:`\exnaddr`.

.. math::
   \begin{array}{lclll}

>    \F{exception\_alloc}(S, \X{gt}, v) &=& (S', \X{a}) && (\iff \allocglobal(S, \X{gt}, v) = S', \X{a}) \\

Reviewer: ```suggestion
   \F{exception\_alloc}(S, \tagaddr, \val^\ast) &=& (S', a) && (\iff \allocexn(S, \tagaddr, \val^\ast) = S', a) \\
```

   \end{array}

.. _embed-tag-type:

> :math:`\F{tag\_type}(\store, \tagaddr) : \tagtype`
> ........................................................

Reviewer: ```suggestion
:math:`\F{tag\_type}(\store, \tagaddr) : \tagtype`
..................................................
```

   \end{array}

.. _embed-tag-type:

:math:`\F{tag\_type}(\store, \tagaddr) : \tagtype`
........................................................

1. Return :math:`S.\STAGS[a].\TAGITYPE`.

2. Post-condition: the returned :ref:`tag type <syntax-tagtype>` is :ref:`valid  <valid-tagtype>`.

.. math::
   \begin{array}{lclll}
   \F{tag\_type}(S, a) &=& S.\STAGS[a].\TAGITYPE \\
   \end{array}

.. index:: exception, exception address, store, exception instance, exception type
.. _embed-exception:

> Exceptions

Reviewer: I think we also need an exception_read to access the exn arguments.

Perhaps shorten names to exn_alloc/exn_read, for consistency with other naming?

Exceptions
~~~~~~~~~~

.. _embed-exn-alloc:

:math:`\F{exn\_alloc}(\store, \tagaddr, \val^\ast) : (\store, \exnaddr)`

> ............................................................................

Reviewer: Nit: adjust underline

   \end{array}

.. _embed-exn-read:

:math:`\F{exn\_read}(\store, \exnaddr) : (\tagaddr, \val^\ast)`

> ......................................................................

Reviewer: Adjust underline to shut up Sphinx.
